src/Vault.sol

```diff
@@ -391,32 +401,48 @@ contract Vault is EIP1967Admin, VaultAccessControl, IERC721Receiver, ICDP, Multi
     /// @param vaultId Id of the vault
     /// @param amount The debt amount to be burned
     function burnDebt(uint256 vaultId, uint256 amount) external {
-        _updateVaultStabilisationFee(vaultId);
-
-        uint256 currentVaultDebt = vaultDebt[vaultId];
-        uint256 overallDebt = stabilisationFeeVaultSnapshot[vaultId] + currentVaultDebt;
-        amount = (amount < overallDebt) ? amount : overallDebt;
-        uint256 overallAmount = amount;
-
-        if (amount > currentVaultDebt) {
-            uint256 burningFeeAmount = amount - currentVaultDebt;
-            token.mint(treasury, burningFeeAmount);
-            stabilisationFeeVaultSnapshot[vaultId] -= burningFeeAmount;
-            amount -= burningFeeAmount;
+        _updateRateFee();
+
+        uint256 overallDebt = getOverallDebt(vaultId);
+        amount = (overallDebt < amount) ? overallDebt : amount;
+
+        token.transferFrom(msg.sender, address(this), amount);
+
+        globalDebtAccumulator -= amount;
+        uint256 tokensToBurn;
+        uint256 currentMintsToBePaid = vaultMintsToBePaid[vaultId];
+
+        if (amount == overallDebt) {
+            tokensToBurn = currentMintsToBePaid;
+            vaultNormalizedDebt[vaultId] = 0;
+            vaultMintsToBePaid[vaultId] = 0;
+        } else {
+            uint256 currentGlobalFeesRateD = globalFeesRateDebtAccumulatorD;
+
+            tokensToBurn = FullMath.mulDiv(currentMintsToBePaid, amount, overallDebt);
+
+            uint256 currentNormalizedDebt = vaultNormalizedDebt[vaultId];
+            uint256 normalizedDebtToBurn = FullMath.mulDiv(amount, DEBT_DENOMINATOR, currentGlobalFeesRateD);
+            if (currentNormalizedDebt < normalizedDebtToBurn) {
+                normalizedDebtToBurn = currentNormalizedDebt;
+            }
+
+            vaultNormalizedDebt[vaultId] -= normalizedDebtToBurn;
+            vaultMintsToBePaid[vaultId] -= tokensToBurn;
         }
 
-        token.transferFrom(msg.sender, address(this), overallAmount);
-        token.burn(overallAmount);
-        vaultDebt[vaultId] -= amount;
+        token.transferAndCall(address(treasury), amount - tokensToBurn, "");
```

Comment:
Consider adding a check on amount == tokensToBurn. In order to avoid unnecessary calls